pkg/ccl/streamingccl/ingestion/stream_ingestion_planning.go

```diff
@@ -0,0 +1,123 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Licensed as a CockroachDB Enterprise file under the Cockroach Community
+// License (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
+
+package ingestion
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/ccl/streamingccl"
+	"github.com/cockroachdb/cockroach/pkg/ccl/utilccl"
+	"github.com/cockroachdb/cockroach/pkg/jobs"
+	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/sql"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/colinfo"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing"
+	"github.com/cockroachdb/errors"
+)
+
+func streamIngestionJobDescription(
+	p sql.PlanHookState, streamIngestion *tree.StreamIngestion,
+) (string, error) {
+	ann := p.ExtendedEvalContext().Annotations
+	return tree.AsStringWithFQNames(streamIngestion, ann), nil
+}
+
+func ingestionPlanHook(
+	ctx context.Context, stmt tree.Statement, p sql.PlanHookState,
+) (sql.PlanHookRowFn, colinfo.ResultColumns, []sql.PlanNode, bool, error) {
+	ingestionStmt, ok := stmt.(*tree.StreamIngestion)
+	if !ok {
+		return nil, nil, nil, false, nil
+	}
+
+	fromFn, err := p.TypeAsStringArray(ctx, tree.Exprs(ingestionStmt.From), "INGESTION")
+	if err != nil {
+		return nil, nil, nil, false, err
+	}
+
+	fn := func(ctx context.Context, _ []sql.PlanNode, resultsCh chan<- tree.Datums) error {
+		ctx, span := tracing.ChildSpan(ctx, stmt.StatementTag())
+		defer span.Finish()
+
+		if err := utilccl.CheckEnterpriseEnabled(
+			p.ExecCfg().Settings, p.ExecCfg().ClusterID(), p.ExecCfg().Organization(),
+			"RESTORE FROM REPLICATION STREAM",
+		); err != nil {
+			return err
+		}
+
+		from, err := fromFn()
+		if err != nil {
+			return err
+		}
+
+		// We only support a TENANT target, so error out if that is nil.
+		if ingestionStmt.Targets.Tenant == (roachpb.TenantID{}) {
+			return errors.Newf("no tenant specified in ingestion query: %s", ingestionStmt.String())
+		}
+
+		// Add privileges checks. Probably the same as RESTORE.
+
+		prefix := keys.MakeTenantPrefix(ingestionStmt.Targets.Tenant)
+		streamIngestionDetails := jobspb.StreamIngestionDetails{
+			StreamAddress: streamingccl.StreamAddress(from[0]),
+			Span:          roachpb.Span{Key: prefix, EndKey: prefix.Next()},
+			// TODO: Figure out what the initial ts should be.
+			StartTime: hlc.Timestamp{},
```

Comment:
This will end up being the `endtime` of the backup that we're restoring when that is hooked up. I think that keeping it at empty should be good for now.